### ISSUE TITLE
nolibc error?

### ISSUE THREAD

**Issue by YoSTEALTH** [axboe/liburing#475]
Tag @ammarfaizi2 Just updated kernel to `Linux 5.15` and pulled current master, updated defines:
``` Bash
prefix                        /usr
includedir                    /usr/include
libdir                        /usr/lib
libdevdir                     /usr/lib
relativelibdir                
mandir                        /usr/man
datadir                       /usr/share
stringop_overflow             yes
array_bounds                  yes
__kernel_rwf_t                yes
__kernel_timespec             yes
open_how                      yes
statx                         yes
C++                           yes
has_ucontext                  yes
has_memfd_create              yes
liburing_nolibc               no
CC                            gcc
CXX                           g++
running build_ext
building 'uring._uring' extension
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/queue.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/queue.o -Os -g0
In file included from libs/liburing/src/queue.c:5:                                                                
libs/liburing/src/syscall.h: In function ‘____sys_io_uring_register’:                                             
libs/liburing/src/syscall.h:101:15: warning: implicit declaration of function ‘syscall’ [-Wimplicit-function-declaration]                                                                                                           
  101 |         ret = syscall(__NR_io_uring_register, fd, opcode, arg, nr_args);                                  
      |               ^~~~~~~                                                                                     
libs/liburing/src/syscall.h: In function ‘uring_madvise’:                                                         
libs/liburing/src/syscall.h:170:15: warning: implicit declaration of function ‘madvise’; did you mean ‘raise’? [-Wimplicit-function-declaration]                                                                                    
  170 |         ret = madvise(addr, length, advice);                                                              
      |               ^~~~~~~                                                                                     
      |               raise                                                                                       
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/register.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/register.o -Os -g0
In file included from libs/liburing/src/register.c:5:                                                             
libs/liburing/src/syscall.h: In function ‘____sys_io_uring_register’:                                             
libs/liburing/src/syscall.h:101:15: warning: implicit declaration of function ‘syscall’ [-Wimplicit-function-declaration]                                                                                                           
  101 |         ret = syscall(__NR_io_uring_register, fd, opcode, arg, nr_args);                                  
      |               ^~~~~~~                                                                                     
libs/liburing/src/syscall.h: In function ‘uring_madvise’:                                                         
libs/liburing/src/syscall.h:170:15: warning: implicit declaration of function ‘madvise’; did you mean ‘raise’? [-Wimplicit-function-declaration]                                                                                    
  170 |         ret = madvise(addr, length, advice);                                                              
      |               ^~~~~~~                                                                                     
      |               raise                                                                                       
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/setup.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/setup.o -Os -g0
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/syscall.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/syscall.o -Os -g0
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c src/uring.c -o build/temp.linux-x86_64-3.10/src/uring.o -Os -g0
gcc -pthread -shared build/temp.linux-x86_64-3.10/libs/liburing/src/queue.o build/temp.linux-x86_64-3.10/libs/liburing/src/register.o build/temp.linux-x86_64-3.10/libs/liburing/src/setup.o build/temp.linux-x86_64-3.10/libs/liburing/src/syscall.o build/temp.linux-x86_64-3.10/src/uring.o -o build/lib.linux-x86_64-3.10/uring/_uring.cpython-310-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-3.10/uring/_uring.cpython-310-x86_64-linux-gnu.so -> uring
running clean
removing 'build/temp.linux-x86_64-3.10' (and everything under it)
removing 'build/lib.linux-x86_64-3.10' (and everything under it)
removing 'build/bdist.linux-x86_64' (and everything under it)
'build/scripts-3.10' does not exist -- can't clean it
removing 'build'
```

**Comment by ammarfaizi2** [comment]
How did you build this? Can you show the full commands you executed?

Your configure output doesn't show that you're building with nolibc option.
```
liburing_nolibc               no
```

The nolibc build works just fine on the current master branch, see the build bot there, it's all green.

To build liburing nolibc, you can just simply execute these:
```
./configure --nolibc;
make -j$(nproc);
```

**Comment by YoSTEALTH** [comment]
I have tired it with the `--nolibc`, I get the same error. Doesn't matter if `--nolibc` is set or not `liburing_nolibc no` is same.

**Comment by ammarfaizi2** [comment]
It seems you've modified the build script?

From what I can see here, you don't have this flag in your build:
```
-include ../config-host.h
```

It's where the `CONFIG_NOLIBC` is defined.

**Comment by ammarfaizi2** [comment]
Works just fine here. Consult your build modification needs for further assistance.

```
ammarfaizi2@integral:~/project/now/liburing$ ./configure --nolibc 
prefix                        /usr
includedir                    /usr/include
libdir                        /usr/lib
libdevdir                     /usr/lib
relativelibdir                
mandir                        /usr/man
datadir                       /usr/share
stringop_overflow             yes
array_bounds                  yes
__kernel_rwf_t                yes
__kernel_timespec             yes
open_how                      yes
statx                         yes
C++                           yes
has_ucontext                  yes
has_memfd_create              yes
liburing_nolibc               yes
CC                            gcc
CXX                           g++
ammarfaizi2@integral:~/project/now/liburing$ make -j$(nproc) V=1
make[1]: Entering directory '/home/ammarfaizi2/project/now/liburing/src'
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o setup.ol setup.c
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o queue.ol queue.c
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o register.ol register.c
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o nolibc.ol nolibc.c
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -fPIC -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o setup.os setup.c
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -fPIC -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o queue.os queue.c
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -fPIC -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o register.os register.c
gcc -D_GNU_SOURCE -Iinclude/ -include ../config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -fPIC -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -c -o nolibc.os nolibc.c
ar r liburing.a setup.ol queue.ol register.ol nolibc.ol
gcc -fPIC -g -fomit-frame-pointer -O2 -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector -shared -Wl,--version-script=liburing.map -Wl,-soname=liburing.so.2 -o liburing.so.2.1.0 setup.os queue.os register.os nolibc.os  -nostdlib -nodefaultlibs
ar: creating liburing.a
ranlib liburing.a
make[1]: Leaving directory '/home/ammarfaizi2/project/now/liburing/src'

```

**Comment by YoSTEALTH** [comment]
I take Liburing source and custom build `.so` to be used by python.
`python3 setup.py build_ext --inplace` (for testing)

Project is linked to whats inside `liburing/src/` even the `compat.h` file is auto created inside `liburing/src/include/liburing`

This is not the project I am testing with atm but you can visit https://github.com/YoSTEALTH/Liburing/blob/master/builder.py see how its normally done.

**Comment by ammarfaizi2** [comment]
So you create your own build script here?
https://github.com/YoSTEALTH/Liburing/blob/master/builder.py#L44-L49

Some notes for you:
1. For nolibc build, you should not compile the `syscall.c`, instead, you need to compile `nolibc.c`.

2. You must define `CONFIG_NOLIBC` macro in all translation units. It can simply be done by adding this to your compile flags:
```
-include ../config-host.h
```
The file `../config-host.h` is generated by the `configure` script.

3. Make sure your build flags contain all necessary options for nolibc support, such as:
```
-nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector
```

**Comment by YoSTEALTH** [comment]
All the include `*.h` files should be in https://github.com/YoSTEALTH/Liburing/blob/master/builder.py#L49 folder! Thus its not picking up the header file.

**Comment by ammarfaizi2** [comment]
> All the include `*.h` files should be in https://github.com/YoSTEALTH/Liburing/blob/master/builder.py#L49 folder! Thus its not picking up the header file.

You can add
```
-DCONFIG_NOLIBC
```
to your build flags if you don't want to include `config-host.h` though, but the `./configure` script is useless then.

**Comment by YoSTEALTH** [comment]
If I add `-DCONFIG_NOLIBC`
``` Bash
prefix                        /usr
includedir                    /usr/include
libdir                        /usr/lib
libdevdir                     /usr/lib
relativelibdir                
mandir                        /usr/man
datadir                       /usr/share
stringop_overflow             yes
array_bounds                  yes
__kernel_rwf_t                yes
__kernel_timespec             yes
open_how                      yes
statx                         yes
C++                           yes
has_ucontext                  yes
has_memfd_create              yes
liburing_nolibc               yes
CC                            gcc
CXX                           g++
running build_ext
building 'uring._uring' extension
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/queue.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/queue.o -Os -g0 -DCONFIG_NOLIBC
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/register.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/register.o -Os -g0 -DCONFIG_NOLIBC
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/setup.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/setup.o -Os -g0 -DCONFIG_NOLIBC
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/syscall.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/syscall.o -Os -g0 -DCONFIG_NOLIBC
libs/liburing/src/syscall.c:11:3: error: #error "This file should only be compiled for libc build, or for liburing tests"
   11 | # error "This file should only be compiled for libc build, or for liburing tests"                        
      |   ^~~~~                                                                                                  
error: command '/usr/bin/gcc' failed with exit code 1
```

**Comment by ammarfaizi2** [comment]
Read my notes, you no (1) and (3)
https://github.com/axboe/liburing/issues/475#issuecomment-962252683

**Comment by YoSTEALTH** [comment]
If I use `nolibc.c` with `-DCONFIG_NOLIBC` it does work! Just seems sloppy.

**Comment by ammarfaizi2** [comment]
All of these should be handled by default by the liburing build script, it is just you that don't use the provided script.

Actually you can just add extra envars like `CFLAGS` to modify the compilation flag.

**Comment by YoSTEALTH** [comment]
What do you think about:
1. moving all the `*.h` files into `liburing/src/include/liburing` folder?
2. `syscall.*` and `nolibc.*` files has if CONFIG_NOLIBC enable code else hide code?

I am just thinking what type of config users system will have, I would rather not hard code cflags if i can avoid it.

**Comment by ammarfaizi2** [comment]
> What do you think about:
> 
>     1. moving all the `*.h` files into `liburing/src/include/liburing` folder?

I don't think `config-host.h` should be included from the userland code, it's just a generated header file that is only useful for building liburing. The userland should rely on `liburing.h` header.

We don't use `#include "config-host.h"` for that anyway, we do that include from the compilation flags.

>     2. `syscall.*` and `nolibc.*` files has if CONFIG_NOLIBC enable code else hide code?

I don't understand the no (2). Can you rephrase that sentence?

> P.s `nolibc` is confusing! `nolibc=yes|no`, maybe `libc=yes|no`?

There is no `yes|no` value for that. By default we use libc, if we specify `--nolibc`, this means we don't use libc.

**Comment by YoSTEALTH** [comment]
Ok, don't worry about my comments. Thanks @ammarfaizi2 for your help resolving this issue.

**Comment by YoSTEALTH** [comment]
@ammarfaizi2 Actually, this time I am including `config-host.h` so want to get both libc/nolibc right, if I can.

`setup.py` file
```Python
from subprocess import run
from setuptools import Extension, find_packages, setup
from Cython.Build import cythonize


LIBC = False
PACKAGE = 'uring'
SOURCES = ['src/uring.pyx',
           'libs/liburing/src/queue.c',
           'libs/liburing/src/register.c',
           'libs/liburing/src/setup.c']
CONFIGURE = ['./configure']
COMPILE_ARGS = ['-Os', '-g0',   # optimization
                '-include', 'libs/liburing/config-host.h']
if LIBC:
    SOURCES.append('libs/liburing/src/syscall.c')
else:
    SOURCES.append('libs/liburing/src/nolibc.c')
    CONFIGURE.append('--nolibc')
    # COMPILE_ARGS.extend(('-nostdlib', '-nodefaultlibs', '-ffreestanding', '-fno-stack-protector'))


# configure liburing
run(CONFIGURE, cwd='libs/liburing')
# Note: configure is only run so all the check and required files are created.
# Note: setup process does the actual compiling part.
setup(name=PACKAGE,
      packages=find_packages(),
      ext_modules=cythonize([Extension(name=f'{PACKAGE}._{PACKAGE}',  # where the `.so` will be saved.
                                       sources=SOURCES,
                                       include_dirs=['libs/liburing/src/include'],
                                       extra_compile_args=COMPILE_ARGS)],
                            annotate=True,          # for debugging only
                            language_level='3'),    # complains even though using cython 3
      zip_safe=False)

```

`LIBC=False` works good.
``` Bash
prefix                        /usr
includedir                    /usr/include
libdir                        /usr/lib
libdevdir                     /usr/lib
relativelibdir                
mandir                        /usr/man
datadir                       /usr/share
stringop_overflow             yes
array_bounds                  yes
__kernel_rwf_t                yes
__kernel_timespec             yes
open_how                      yes
statx                         yes
C++                           yes
has_ucontext                  yes
has_memfd_create              yes
liburing_nolibc               yes
CC                            gcc
CXX                           g++
running build_ext
building 'uring._uring' extension
creating build
creating build/temp.linux-x86_64-3.10
creating build/temp.linux-x86_64-3.10/libs
creating build/temp.linux-x86_64-3.10/libs/liburing
creating build/temp.linux-x86_64-3.10/libs/liburing/src
creating build/temp.linux-x86_64-3.10/src
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/nolibc.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/nolibc.o -Os -g0 -include libs/liburing/config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/queue.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/queue.o -Os -g0 -include libs/liburing/config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/register.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/register.o -Os -g0 -include libs/liburing/config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/setup.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/setup.o -Os -g0 -include libs/liburing/config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c src/uring.c -o build/temp.linux-x86_64-3.10/src/uring.o -Os -g0 -include libs/liburing/config-host.h -nostdlib -nodefaultlibs -ffreestanding -fno-stack-protector
creating build/lib.linux-x86_64-3.10
creating build/lib.linux-x86_64-3.10/uring
gcc -pthread -shared build/temp.linux-x86_64-3.10/libs/liburing/src/nolibc.o build/temp.linux-x86_64-3.10/libs/liburing/src/queue.o build/temp.linux-x86_64-3.10/libs/liburing/src/register.o build/temp.linux-x86_64-3.10/libs/liburing/src/setup.o build/temp.linux-x86_64-3.10/src/uring.o -o build/lib.linux-x86_64-3.10/uring/_uring.cpython-310-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-3.10/uring/_uring.cpython-310-x86_64-linux-gnu.so -> uring
running clean
removing 'build/temp.linux-x86_64-3.10' (and everything under it)
removing 'build/lib.linux-x86_64-3.10' (and everything under it)
'build/bdist.linux-x86_64' does not exist -- can't clean it
'build/scripts-3.10' does not exist -- can't clean it
removing 'build'
```

`LIBC=True` back to error.
``` Bash
prefix                        /usr
includedir                    /usr/include
libdir                        /usr/lib
libdevdir                     /usr/lib
relativelibdir                
mandir                        /usr/man
datadir                       /usr/share
stringop_overflow             yes
array_bounds                  yes
__kernel_rwf_t                yes
__kernel_timespec             yes
open_how                      yes
statx                         yes
C++                           yes
has_ucontext                  yes
has_memfd_create              yes
liburing_nolibc               no
CC                            gcc
CXX                           g++
running build_ext
building 'uring._uring' extension
creating build
creating build/temp.linux-x86_64-3.10
creating build/temp.linux-x86_64-3.10/libs
creating build/temp.linux-x86_64-3.10/libs/liburing
creating build/temp.linux-x86_64-3.10/libs/liburing/src
creating build/temp.linux-x86_64-3.10/src
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/queue.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/queue.o -Os -g0 -include libs/liburing/config-host.h
In file included from libs/liburing/src/queue.c:5:                                                               
libs/liburing/src/syscall.h: In function ‘____sys_io_uring_register’:                                            
libs/liburing/src/syscall.h:101:15: warning: implicit declaration of function ‘syscall’ [-Wimplicit-function-declaration]                                                                                                         
  101 |         ret = syscall(__NR_io_uring_register, fd, opcode, arg, nr_args);                                 
      |               ^~~~~~~                                                                                    
libs/liburing/src/syscall.h: In function ‘uring_madvise’:                                                        
libs/liburing/src/syscall.h:170:15: warning: implicit declaration of function ‘madvise’; did you mean ‘raise’? [-Wimplicit-function-declaration]                                                                                  
  170 |         ret = madvise(addr, length, advice);                                                             
      |               ^~~~~~~                                                                                    
      |               raise                                                                                      
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/register.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/register.o -Os -g0 -include libs/liburing/config-host.h
In file included from libs/liburing/src/register.c:5:                                                            
libs/liburing/src/syscall.h: In function ‘____sys_io_uring_register’:                                            
libs/liburing/src/syscall.h:101:15: warning: implicit declaration of function ‘syscall’ [-Wimplicit-function-declaration]                                                                                                         
  101 |         ret = syscall(__NR_io_uring_register, fd, opcode, arg, nr_args);                                 
      |               ^~~~~~~                                                                                    
libs/liburing/src/syscall.h: In function ‘uring_madvise’:                                                        
libs/liburing/src/syscall.h:170:15: warning: implicit declaration of function ‘madvise’; did you mean ‘raise’? [-Wimplicit-function-declaration]                                                                                  
  170 |         ret = madvise(addr, length, advice);                                                             
      |               ^~~~~~~                                                                                    
      |               raise                                                                                      
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/setup.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/setup.o -Os -g0 -include libs/liburing/config-host.h
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c libs/liburing/src/syscall.c -o build/temp.linux-x86_64-3.10/libs/liburing/src/syscall.o -Os -g0 -include libs/liburing/config-host.h
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Ilibs/liburing/src/include -I/opt/python/3.10/include/python3.10 -c src/uring.c -o build/temp.linux-x86_64-3.10/src/uring.o -Os -g0 -include libs/liburing/config-host.h
creating build/lib.linux-x86_64-3.10
creating build/lib.linux-x86_64-3.10/uring
gcc -pthread -shared build/temp.linux-x86_64-3.10/libs/liburing/src/queue.o build/temp.linux-x86_64-3.10/libs/liburing/src/register.o build/temp.linux-x86_64-3.10/libs/liburing/src/setup.o build/temp.linux-x86_64-3.10/libs/liburing/src/syscall.o build/temp.linux-x86_64-3.10/src/uring.o -o build/lib.linux-x86_64-3.10/uring/_uring.cpython-310-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-3.10/uring/_uring.cpython-310-x86_64-linux-gnu.so -> uring
running clean
removing 'build/temp.linux-x86_64-3.10' (and everything under it)
removing 'build/lib.linux-x86_64-3.10' (and everything under it)
'build/bdist.linux-x86_64' does not exist -- can't clean it
'build/scripts-3.10' does not exist -- can't clean it
removing 'build'
```

**Comment by ammarfaizi2** [comment]
You're missing `-D_GNU_SOURCE`.

**Comment by YoSTEALTH** [comment]
That did it. Thanks @ammarfaizi2 works like a charm :)
